bftengine/src/bcstatetransfer/RVBManager.cpp

```diff
@@ -0,0 +1,708 @@
+  // data and the removal is reflected in the RVB persisted tree of the last checkpoint.
+  //
+  // We must prune the pruning vector based on some persisted checkpoint in the past.
+  // We cannot allow losing digests, we will never get them back (blocks are not in storage anymore) and will have
+  // to reconstruct the whole tree.
+  if (last_checkpoint_desc_.checkpointNum > 0) {
+    for (i = 0; i < pruned_blocks_digests_.size(); ++i) {
+      if (pruned_blocks_digests_[i].first >= last_checkpoint_desc_.maxBlockId) {
+        break;
+      }
+    }
+    if (i > 0) {
+      LOG_DEBUG(logger_,
+                "Remove " << i << " digests from pruned_blocks_digests_, from/to block IDs:"
+                          << KVLOG(pruned_blocks_digests_[0].first, pruned_blocks_digests_[i].first));
+      pruned_blocks_digests_.erase(pruned_blocks_digests_.begin(), pruned_blocks_digests_.begin() + i);
+      ds_->setPrunedBlocksDigests(pruned_blocks_digests_);
+    }
+  }
+
+  // Second, prune in_mem_rvt_ up to new_checkpoint_desc.maxBlockId.
+  // Theoratically, there might be block digests which belong to the next checkpoint
+  //
+  // Comment: assume RVT span on the range [160 ... 1500] min/max RVB IDs respectively. FetchRangeSize=16 and
+  // new_checkpoint_desc.maxBlockId is 1500.
+  // We might have in the pruning vector the next digests 144, 160 .... 1488, 1504 ...
+  // 144 and 1504 should not be pruned. 144 triggers skip (continue). 1504 triggers break
+  BlockId from_block_id{}, to_block_id{};
+  for (i = 0; i < pruned_blocks_digests_.size(); ++i) {
+    RVBId rvb_id = pruned_blocks_digests_[i].first;
+    auto digest = pruned_blocks_digests_[i].second;
+    if ((rvb_id <= new_checkpoint_desc.maxBlockId) && (rvb_id == in_mem_rvt_->getMinRvbId())) {
+      LOG_TRACE(logger_, "Remove digest for block " << rvb_id << " ,Digest: " << digest.toString());
+      if (!from_block_id) from_block_id = rvb_id;
+      to_block_id = rvb_id;
+      in_mem_rvt_->removeNode(rvb_id, digest.get(), BLOCK_DIGEST_SIZE);
+    } else if (rvb_id > new_checkpoint_desc.maxBlockId) {
+      break;
+    }
+  }
+  if (i > 0) {
+    LOG_INFO(
+        logger_,
+        "Removed " << (i - 1) << " digests from in_mem_rvt_, from/to block IDs:" << KVLOG(from_block_id, to_block_id));
+  }
+  if (!pruned_blocks_digests_.empty() && (debug_prints_log_level.find(getLogLevel()) != debug_prints_log_level.end())) {
+    ostringstream oss;
+    oss << "pruned_blocks_digests_: size: " << pruned_blocks_digests_.size() << " Pairs: ";
+    for (auto const& pair : pruned_blocks_digests_) {
+      oss << ",[" << std::to_string(pair.first) << "," << pair.second.toString() << "]";
+    }
+    LOG_DEBUG(logger_, oss.str());
+  }
+  // We relay on caller to persist new_checkpoint_desc, and leave pruned_blocks_digests_ persisted before erase
+  // was done (some redundent digests)
+}
+
+void RVBManager::updateRvbDataDuringCheckpoint(CheckpointDesc& new_checkpoint_desc) {
```

Comment:
const reference

Reply:
not here, we want to use it as in and out.